src/common/ad_mem.c

```diff
@@ -3,5 +3,1917 @@
+	group_decref(grp);
+	return rc;
+}
+
+/**
+ * Reserve storage space with specified @size, the space should be allocated from
+ * default arena if @arena_id is set to zero, otherwise it is allocated from the
+ * provided arena.
+ */
+daos_off_t
+blob_reserve_addr(struct ad_blob *blob, int type, daos_size_t size,
+		  uint32_t *arena_id, struct ad_reserv_act *act)
+{
+	struct ad_arena	*arena = NULL;
+	daos_off_t	 addr;
+	uint32_t	 id;
+	bool		 tried;
+	int		 rc;
+
+	if (arena_id && *arena_id != AD_ARENA_ANY)
+		id = *arena_id;
+	else
+		id = blob->bb_arena_last[type]; /* the last used arena */
+
+	/* arena zero should have been created while initializing the blob */
+	D_ASSERT(id != AD_ARENA_ANY);
+	D_DEBUG(DB_TRACE, "Loading arena=%u\n", id);
+
+	rc = arena_load(blob, id, &arena);
+	if (rc) {
+		D_DEBUG(DB_TRACE, "Failed to load arena %u: %d\n", id, rc);
+		/* fall through and create a new one */
+
+	} else if (arena->ar_full) {
+		D_DEBUG(DB_TRACE, "Arena %u is full, create a new one\n", id);
+		arena_decref(arena);
+		arena = NULL;
+	}
+
+	tried = false;
+	while (1) {
+		if (arena == NULL) {
+			rc = arena_reserve(blob, type, NULL, &arena);
+			if (rc) {
+				D_ERROR("Failed to reserve new arena: %d\n", rc);
+				return rc;
+			}
+			tried = true;
+		}
+
+		D_DEBUG(DB_TRACE, "reserve from arena==%d\n", arena->ar_df->ad_id);
+		addr = arena_reserve_addr(arena, size, act);
+		if (!addr) { /* full arena */
+			D_DEBUG(DB_TRACE, "full arena=%d\n", arena->ar_df->ad_id);
+			arena_decref(arena);
+			if (tried)
+				return 0;
+
+			arena = NULL;
+			continue;
+		}
+
+		/* completed */
+		if (arena_id)
+			*arena_id = arena->ar_df->ad_id;
+
+		arena_decref(arena);
+		return addr;
+	}
+}
+
+daos_off_t
+ad_reserve(struct ad_blob_handle bh, int type, daos_size_t size,
+	   uint32_t *arena_id, struct ad_reserv_act *act)
+{
+	return blob_reserve_addr(bh.bh_blob, type, size, arena_id, act);
+}
+
+int
+tx_complete(struct ad_tx *tx, int err)
+{
+	struct ad_blob	   *blob  = tx->tx_blob;
+	struct umem_store  *store = &blob->bb_store;
+	struct ad_arena    *arena;
+	struct ad_group	   *group;
+	struct ad_free_act *fac;
+	int		    i;
+	int		    rc;
+
+	if (!err)
+		rc = store->stor_ops->so_wal_submit(store, tx->tx_id, &tx->tx_redo);
+	else
+		rc = err;
+
+	while ((arena = d_list_pop_entry(&tx->tx_ar_pub, struct ad_arena, ar_link))) {
+		arena->ar_publishing = false;
+		if (err) { /* keep the refcount and pin it */
+			d_list_add(&arena->ar_link, &blob->bb_ars_rsv);
+			continue;
+		}
+		clrbit(blob->bb_bmap_rsv, arena->ar_df->ad_id);
+		D_ASSERT(arena->ar_unpub);
+		arena->ar_unpub = false;
+		arena_decref(arena);
+	}
+
+	while ((group = d_list_pop_entry(&tx->tx_gp_pub, struct ad_group, gp_link))) {
+		group->gp_publishing = false;
+		if (err) { /* keep the refcount and pin it */
+			d_list_add(&group->gp_link, &blob->bb_gps_rsv);
+			continue;
+		}
+		for (i = 0; i < group->gp_bit_nr; i++)
+			clrbit(group->gp_arena->ar_bmap_rsv, group->gp_bit_at + i);
```

Comment:
Same here. Isn't the reserve bit cleared (turned into allocation bits) on publish?

Reply:
This is called on successful committed transaction, if the transaction is canceld/aborted, we don't reset these bits.

Reply:
I mean could the same reserve bit being set by another tx while this tx yield on wal commit?

Reply:
seems cannot reserve/set by another tx again, as the "ad->ad_bmap" is already set?

Reply:
yeah, seems so. Then I'm wondering what's the purpose of reserve bit in such case? I can understand reserve bit for a particular allocation, but I don't quite see why do we need reserve bit arena or group? Arena & group are shared by transactions anyway, right?